### PR TITLE
Added similar-string custom pylint linter

### LINKED ISSUE #1126: new linter: [l10n] reduce the number of similar strings
There are some very similar (sometimes even the same) strings. They differ in capitalization or spelling.  In general we'd like to minimize the number of translatable strings to make it easier for language translators.

Kiwi TCMS needs a custom kiwi_lint/ checker which will discover all similar strings (that are marked for translation) and warn about them so we can gradually remove duplicates.

See the `textdistance` package mentioned in the comment below. That sounds like a good candidate to plug into our customized pylint infrastructure.

**Definition of Done**:
- new checker class under `kiwi_lint/` which implements the necessary pylint machinery
- Checker uses the levenstein distance algorithm (for starters), we may decide to change it later.
- NOTE: initially the checker probably wants to pay attention only to strings marked for translation (as opposed to any kind of string constant), see #733 
- NOTE2: at some point before this checker is merged we need to teach it how to parse templates and extract all strings marked for translation before comparing them
- Note3: this checker will likely need to consider "similarity percentage". Let's say 80% or above would be considered duplicate strings. The actual value must be a class attribute so we can play around with it, later we may turn it into a linter configuration.
- All code review passed & PR merged

Here's me explaining how to write pylint plugins:
- https://www.youtube.com/watch?v=3CkSKUNMLJc
- https://www.youtube.com/watch?v=D1wLFyazTRM

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by 17sushmita**
Fixes #1126 

- Added new linter- `similar-string` to detect similar strings which are marked for translation i.e strings passed to _() or gettext_lazy()
- Used Levenshtein distance from textdistance module which produced quite decent results. You might check if we need more complex similar strings to be detected, I'll use other algorithms for text distance.

### CODE REVIEW

**Comment by 17sushmita** [comment]
@atodorov , Where should I place the dependency `textdistance` in requirements/ ? The tests are failing wherever I place it.

**Comment by atodorov** [comment]
> @atodorov , Where should I place the dependency `textdistance` in requirements/ ? The tests are failing wherever I place it.

This should be in `devel.txt`

File: kiwi_lint/similar_string.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    @staticmethod
+    def clean_string(text):
+        for word in text:
+            if word not in string.punctuation:
+                text += "".join(word)
+        return text
+
+    @staticmethod
+    def marked_for_translation(funcname):
+        if funcname in ("_", "gettext_lazy"):
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:34]: the if condition is a boolean expression in itself so no need for the if-return construct.

File: kiwi_lint/similar_string.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            return True
+        return False
+
+    def check_similar_string(self, translation_string):
+        for key in self._dict_of_strings:
+            if (
+                levenshtein.normalized_similarity(
+                    self.clean_string(translation_string), self.clean_string(key)
+                )
+                >= 0.8
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:44]: make the percentage a class attribute according to issue description.

File: kiwi_lint/similar_string.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            ):
+                return key
+        return None
+
+    def check_same_string(self, string):
+        if string in self._dict_of_strings:
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:50]: Same comment for if/return like above.

File: kiwi_lint/similar_string.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    def __init__(self, linter):
+        super().__init__(linter)
+        self._dict_of_strings = {}
+
+    @staticmethod
+    def clean_string(text):
+        for word in text:
+            if word not in string.punctuation:
+                text += "".join(word)
+        return text
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:30]: Please explain the idea behind this method? Why do we need to clean up the strings ? Isn't the Levenshtein algorithm going to produce similar results if the strings aren't cleaned ?

File: requirements/base.txt [hunk 5]
```diff
@@ -16,4 +16,4 @@ Markdown==3.3.4
 Pygments==2.8.1
 python-gitlab==2.6.0
 python-redmine==2.3.0
-topicaxis-opengraph==0.5
+topicaxis-opengraph==0.5
```
Comment (atodorov) [inline requirements/base.txt:19]: Not sure what changed here but it definitely unrelated. Revert the change.

File: requirements/readthedocs.txt [hunk 6]
```diff
@@ -11,4 +11,4 @@ pydot
 # linting tools
 pylint-django
 twine
-pyOpenSSL
+pyOpenSSL
```
Comment (atodorov) [inline requirements/readthedocs.txt:14]: Not sure what changed here but it definitely unrelated. Revert the change.

**Comment by atodorov** [comment]
```
ERROR: kiwi_lint (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: kiwi_lint
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/unittest/loader.py", line 470, in _find_test_path
    package = self._get_module_from_name(name)
  File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/home/runner/work/Kiwi/Kiwi/kiwi_lint/__init__.py", line 27, in <module>
    from .similar_string import SimilarStringChecker
  File "/home/runner/work/Kiwi/Kiwi/kiwi_lint/similar_string.py", line 6, in <module>
    from textdistance import levenshtein
ModuleNotFoundError: No module named 'textdistance'
```

This is strange. It looks like the unittest discovery mechanism is trying to load the linter modules and possibly run tests from them but because there's a missing dependency it fails. There are no tests in this directory at all so this isn't even supposed to happen. IDK why that is.

File: kiwi_lint/similar_string.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            if (
+                levenshtein.normalized_similarity(
+                    self.clean_string(translation_string), self.clean_string(key)
+                )
+                >= 0.8
+            ):
+                return key
+        return None
+
+    def check_same_string(self, string):
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:49]: see pylint log.

**Review by atodorov — CHANGES_REQUESTED** [review]
Changes requested.

Also add the issue number reference in the commit log.

**Comment by 17sushmita** [comment]
> ```
> ERROR: kiwi_lint (unittest.loader._FailedTest)
> ----------------------------------------------------------------------
> ImportError: Failed to import test module: kiwi_lint
> Traceback (most recent call last):
>   File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/unittest/loader.py", line 470, in _find_test_path
>     package = self._get_module_from_name(name)
>   File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/unittest/loader.py", line 377, in _get_module_from_name
>     __import__(name)
>   File "/home/runner/work/Kiwi/Kiwi/kiwi_lint/__init__.py", line 27, in <module>
>     from .similar_string import SimilarStringChecker
>   File "/home/runner/work/Kiwi/Kiwi/kiwi_lint/similar_string.py", line 6, in <module>
>     from textdistance import levenshtein
> ModuleNotFoundError: No module named 'textdistance'
> ```
> 
> This is strange. It looks like the unittest discovery mechanism is trying to load the linter modules and possibly run tests from them but because there's a missing dependency it fails. There are no tests in this directory at all so this isn't even supposed to happen. IDK why that is.

Yes, I tried to add this dependency in devel.txt before and in other requirements file too but this error doesn't seem to resolve

File: requirements/readthedocs.txt (hunk 6 above)
Comment (17sushmita) [reply requirements/readthedocs.txt:14]: Actually I tried to add textdistance module here but removed later in next push maybe that's why it is showing to be changed

File: kiwi_lint/similar_string.py (hunk 4 above)
Comment (17sushmita) [reply kiwi_lint/similar_string.py:30]: I created this method to clean the punctuations in strings. Without using this function there are many false-positive results which includes 
```
tcms/signals.py:222:16: R5011: A similar string to "NEW: Bug #%(pk)d - %(summary)s" exists: "NEW: TestRun #%(pk)d - %(summary)s" (similar-string)
```
here the operators used in this string is adding up to the similarity of the two string which is not required to be detected.

**Comment by 17sushmita** [comment]
@atodorov , This PR is ready for re-review

**Comment by 17sushmita** [comment]
@atodorov , This PR is ready for re-review

**Comment by 17sushmita** [comment]
@atodorov I was wondering if I also need to resolve the linting errors generated by this new linter check?

File: kiwi_lint/similar_string.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+from textdistance import levenshtein
+
+
+class SimilarStringChecker(BaseChecker):
+    __implements__ = (IAstroidChecker,)
+    name = "similar-string-checker"
+
+    msgs = {
+        "R5011": (
+            'A similar string to "%s" exists: "%s"',
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:15]: Let's change the message to `"%s" is similar to "%s". Try keeping them the same to reduce work for translators!`

File: kiwi_lint/similar_string.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            "similar-string",
+            "Similar strings should be avoided for single translation",
+        )
+    }
+
+    def __init__(self, linter):
+        super().__init__(linter)
+        self._dict_of_strings = {}
+        self.threshold = 0.8
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:24]: You can make these class attributes and we'll not need the init method.

File: kiwi_lint/similar_string.py (hunk 4 above)
Comment (atodorov) [reply kiwi_lint/similar_string.py:31]: Got it. Good catch. However put this explanation into the doc-string for the method because this sound important.

File: kiwi_lint/similar_string.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+                levenshtein.normalized_similarity(
+                    self.clean_string(translation_string), self.clean_string(key)
+                )
+                >= self.threshold
+            ):
+                return key
+        return None
+
+    def check_same_string(self, translation_string):
+        return translation_string in self._dict_of_strings
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:49]: method used only in 1 place. Remove the method and use the condition directly. It makes the visit_call method more readable.

File: kiwi_lint/similar_string.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+    @utils.check_messages("similar-string")
+    def visit_call(self, node):
+        if self.marked_for_translation(node.func.as_string()) and isinstance(
+            node.args[0], astroid.nodes.Const
+        ):
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:55]: Instead of nested if conditions I prefer to have an if-return sequence here. It makes the algorithm more easier to follow in case it needs to be changed later.

File: kiwi_lint/similar_string.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            translation_string = node.args[0].value
+            if not self.check_same_string(
+                translation_string
+            ) and self.check_similar_string(translation_string):
+                self.add_message(
+                    "similar-string",
+                    node=node,
+                    args=(
+                        translation_string,
+                        self.check_similar_string(translation_string),
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:65]: avoid double calculation by assigning the value to a variable.

File: kiwi_lint/similar_string.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+    @staticmethod
+    def marked_for_translation(funcname):
+        return funcname in ("_", "gettext_lazy")
+
+    def check_similar_string(self, translation_string):
+        for key in self._dict_of_strings:
+            if (
+                levenshtein.normalized_similarity(
+                    self.clean_string(translation_string), self.clean_string(key)
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:41]: Avoid cleaning `translation_string` on every iteration. The results of this operation don't change so move that before the for loop.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by atodorov** [comment]
> @atodorov I was wondering if I also need to resolve the linting errors generated by this new linter check?

No. However it sounds like a good idea to actually disable this checker when running pylint and execute it as a separate CI job which we will allow to fail for now.

**Comment by atodorov** [comment]
@17sushmita note: please don't mark review comments as resolved. This is the job of the reviewer and it just makes the process of expanding/collapsing these comments more cumbersome.

File: kiwi_lint/similar_string.py (hunk 11 above)
Comment (atodorov) [reply kiwi_lint/similar_string.py:55]:
```
if not self.marked_for_translation():
    return

if not ...
    return


```

**Comment by 17sushmita** [comment]
@atodorov Can you check why code coverage is reducing? I'm not authorized to check.

File: .github/workflows/lint.yml [hunk 14]
```diff
@@ -74,12 +74,31 @@ jobs:
...
     - name: Lint with pylint
       run: |
-        pip install pylint pylint-django
+        pip install pylint pylint-django textdistance
         pip install -r requirements/mariadb.txt
         pip install -r requirements/postgres.txt
 
         make pylint_site_packages
 
+  pylint_similar_string:
```
Comment (atodorov) [inline .github/workflows/lint.yml:83]: Call that similar_strings and adjust the make target accordingly.

File: kiwi_lint/similar_string.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    threshold = 0.8
+
+    @staticmethod
+    def clean_string(text):
+        """
+        This method removes the operators and other punctuations
+        used in the string to avoid unwanted data to add up to
+        the similarity between the strings.
+        """
+        for word in text:
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:31]: this is `character`, not `word`

File: kiwi_lint/similar_string.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+            if word not in string.punctuation:
+                text += "".join(word)
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:33]: You don't need to use `.join()`, concatenation to `text` produces the same result.

Comment (atodorov) [reply kiwi_lint/similar_string.py:33]: However modifying the input parameter while looping over it is not a good idea. Moreover the results are incorrect.

```
<<< input: NEW: Bug #%(pk)d - %(summary)s
>>> output: NEW: Bug #%(pk)d - %(summary)sNEW Bug pkd  summarys
```

File: kiwi_lint/similar_string.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    @utils.check_messages("similar-string")
+    def visit_call(self, node):
+        if not self.marked_for_translation(node.func.as_string()):
+            return
+
+        if not isinstance(node.args[0], astroid.nodes.Const):
+            return
+
+        translation_string = node.args[0].value
+        similar_string = self.check_similar_string(translation_string)
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:61]: `if translation_string in self._dict_of_strings` => return b/c we've seen this string before and it's exactly the same as it was so no need to calculate the similarity at all.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: kiwi_lint/similar_string.py (hunk 16 above)
Comment (17sushmita) [reply kiwi_lint/similar_string.py:33]: should I leave '(' and ')' while checking for punctuation?

File: kiwi_lint/similar_string.py [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+                    cleaned_translation_string, self.clean_string(key)
+                )
+                >= self.threshold
+            ):
+                return key
+        return None
+
+    @utils.check_messages("similar-string")
+    def visit_call(self, node):
+        if not self.marked_for_translation(node.func.as_string()):
```
[the remark this answers is not in the document]
Comment (17sushmita) [reply kiwi_lint/similar_string.py:54]: At some instances, it is not working as expected for example - When the function call `os.environ.setdefault()` is visited, it throws an exception `AttributeError: 'Attribute' object has no attribute 'name'`. We'd need a try catch block to handle this. For simplicity I think the `as_string()` works fine. What do you say?

File: kiwi_lint/similar_string.py (hunk 16 above)
Comment (atodorov) [reply kiwi_lint/similar_string.py:33]: don't change the if condition for now, change how the method collects its result.

File: kiwi_lint/similar_string.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (atodorov) [reply kiwi_lint/similar_string.py]: The exception comes from `.environ` which is an attribute, not a method however both of them can be children of a Call node. See generic_foreign_key.py for example.

`.as_string()` is a hack and I'd prefer a more robust implementation when that isn't overly complicated which in this case it isn't. Python is an explicit language so let the implementation be as explicit as possible.

File: .github/workflows/lint.yml [hunk 19]
```diff
@@ -74,12 +74,31 @@ jobs:
         python-version: ${{ matrix.python-version }}
     - name: Lint with pylint
       run: |
-        pip install pylint pylint-django
+        pip install pylint pylint-django textdistance
```
Comment (atodorov) [inline .github/workflows/lint.yml:77]: I just realized this should be `pip install -r requirements/devel.txt`. I think it has been mistakenly written the first time around. In any case we don't want to be defining dependencies outside of the requirements files.

File: kiwi_lint/similar_string.py [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    @staticmethod
+    def clean_string(text):
+        """
+        This method removes the operators and other punctuations
+        used in the string to avoid unwanted data to add up to
+        the similarity between the strings.
+        """
+        for char in text:
+            if char not in string.punctuation:
+                text += "" + char
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:33]: `+=` means `text = text + `. No need for empty string literal.

File: kiwi_lint/similar_string.py [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+        return text
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:34]: This is still not returning appropriate results. Play with it and you will see.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: kiwi_lint/similar_string.py (hunk 21 above)
Comment (17sushmita) [reply kiwi_lint/similar_string.py:34]: Oh, sorry! I didn't realize this.

**Comment by 17sushmita** [comment]
@atodorov, earlier, I made the marked_for_translation a function because of the code climate issue. The cyclomatic complexity has increased due to adding more conditions in the function. Now, because we're using node.func.name, I need to add another condition. What should I do in that case? Should I make the function again or use node.func.as_string() or if there is any better way to do this?

**Comment by atodorov** [comment]
> @atodorov, earlier, I made the marked_for_translation a function because of the code climate issue. The cyclomatic complexity has increased due to adding more conditions in the function. Now, because we're using node.func.name, I need to add another condition. What should I do in that case? Should I make the function again or use node.func.as_string() or if there is any better way to do this?

Don't pay attention to CodeClimate in this case.

**Comment by 17sushmita** [comment]
> > @atodorov, earlier, I made the marked_for_translation a function because of the code climate issue. The cyclomatic complexity has increased due to adding more conditions in the function. Now, because we're using node.func.name, I need to add another condition. What should I do in that case? Should I make the function again or use node.func.as_string() or if there is any better way to do this?
> 
> Don't pay attention to CodeClimate in this case.

Okay, and Could you check why codecoverage is reducing this much? I'm not authorized to check

**Comment by atodorov** [comment]
> Okay, and Could you check why codecoverage is reducing this much? I'm not authorized to check

Don't bother with that either. It is reducing because you don't have permissions to execute some test jobs and they fail hence less coverage.

File: .github/workflows/lint.yml [hunk 22]
```diff
@@ -77,9 +77,29 @@ jobs:
         pip install pylint pylint-django
```
Comment (atodorov) [inline .github/workflows/lint.yml:77]: If you are installing devel.txt below you don't need this line.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor changes requested. 

Then we're left with `NOTE2` from the original issue description - how to teach this plugin to read the html template files and extract user visible strings from there.

File: kiwi_lint/similar_string.py [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+        html_files_to_lint.extend(
+            os.path.join(dirpath, f) for f in filenames if f.endswith(".html")
+        )
+
+for filepath in html_files_to_lint:
+    with open(filepath, "r") as f:
+        read_lines = f.readlines()
+        file_lines = tuple(read_lines)
+    parser = SimilarStringChecker()
+    parser.parse_translation_string(filepath, file_lines)
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:140]: See open & close methods in empty.py.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by 17sushmita** [comment]
@atodorov, Could you please review this when you're free?

File: kiwi_lint/similar_string.py [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+                    )
+                    continue
+
+                self._dict_of_strings[translation_string] = True
+
+    def close(self):
+        for filepath in self.all_html_files:
+            with open(filepath, "r") as file:
+                read_lines = file.readlines()
+                file_lines = tuple(read_lines)
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:131]: the conversion to tuple doesn't seem to be necessary and the intermeiate variables either. You can pass `file.readlines()` directly to `parse_translation_string()`

File: kiwi_lint/similar_string.py [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+                    similar_string,
+                ),
+            )
+            return
+
+        self._dict_of_strings[translation_string] = True
+
+    def parse_translation_string(self, filename, lines):
+        for lineno, line in enumerate(lines):
+            match_in_line = re.finditer(r'{% trans "([\w ]+)" %}', line)
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:102]: there are also `blocktrans` tags which span multiple lines.

File: kiwi_lint/similar_string.py [hunk 26]
```diff
@@ -0,0 +1,146 @@
...
+    # NOTE: this works against tcms/ directory and will not take into account
+    # if we want to examine only a sub-dir or a few files
+    # all files found by os.walk
+    all_html_files = set()
+
+    def open(self):
+        for rootpath, _dirs, files in os.walk(self.project_root, topdown=False):
+
+            for file_name in files:
+                if file_name.endswith(".html"):
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:40]: There are also .txt files in tcms/templates/email used for email templates.

File: kiwi_lint/similar_string.py [hunk 27]
```diff
@@ -0,0 +1,146 @@
...
+                        (
+                            lineno + 1,
+                            match.start() + 10,
+                            translation_string,
+                            similar_string,
+                        )
+                    )
+                    continue
+
+                self._dict_of_strings[translation_string] = True
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:125]: I don't understand the if-continue from the previous block + this line. Can you walk me through the code here.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: kiwi_lint/similar_string.py (hunk 27 above)
Comment (17sushmita) [reply kiwi_lint/similar_string.py:125]: It is very similar to the if-return from the `visit_call` method, but here we're parsing each string we found in loop instead of function as in previous, so, the first if-continue states that if the same string is found in `_dict_of_strings`, then we should not go to next line instead continue to next iteration(next string) and similarly, second if-continue states that if similar string is found, we add this to `error_messages` and then continue with next string.

Comment (17sushmita) [reply kiwi_lint/similar_string.py:125]: It is very similar to the if-return from the `visit_call` method, in `visit_call`, we were parsing each string in a method call and if same string was found we had to return from the call and continue with the next string. But, here we're parsing each string in a loop and so if same string is found we do not execute next line, instead, we continue to the next iteration i.e. next string. And similarly in the second if-continue, if similar-string is found, we add the message to `error_messages` dict and continue with the next string. And if both the conditions fail, we're left with a distinct string and we add it to `_dict_of_strings`.

**Comment by 17sushmita** [comment]
@atodorov, I've made the requested changes. Please review the PR again.

**Comment by atodorov** [comment]
@17sushmita FYI I haven't forgotten about this but currently super busy with other tasks. I will review probably next week when I have some more free time.

**Comment by 17sushmita** [comment]
> @17sushmita FYI I haven't forgotten about this but currently super busy with other tasks. I will review probably next week when I have some more free time.

No problem, Thanks for informing!😄

File: kiwi_lint/similar_string.py [hunk 28]
```diff
@@ -0,0 +1,190 @@
...
+                    )
+        for rootpath, _dirs, files in os.walk(
+            os.path.join(self.project_root, "templates", "email"), topdown=False
+        ):
+
+            for file_name in files:
+                if file_name.endswith(".txt"):
+                    self.all_template_files.add(
+                        os.path.join(self.project_root, rootpath, file_name)
+                    )
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:52]: Is there a reason for the txt file scanning to be inside a second for loop instead of just moving the if block above under the original loop ?

File: kiwi_lint/similar_string.py [hunk 29]
```diff
@@ -0,0 +1,190 @@
...
+            isinstance(node.func, astroid.Name)
+            and node.func.name in ("_", "gettext_lazy")
+        ):
+            return
+        if not isinstance(node.args[0], astroid.nodes.Const):
+            return
+
+        translation_string = node.args[0].value
+
+        if translation_string in self._dict_of_strings:
```
Comment (atodorov) [inline kiwi_lint/similar_string.py:91]: From here till the end of this method we're doing essentially the same as `check_similar_and_add_error_message` with minor differences in how the error message is annotated.

For the sake of brevity and reusing the same code you can adapt `check_similar_and_add_error_message` and/or `self.error_messages` to work both with a node/filename (or construct the node from filename inside this method) and make the `close()` method simply iterate over `self.error_messages`. Because you have control over the data structure you can even format it like 
```
{
'node': ...,
'line': ...,
'col_offset': ...
'args': (.....)
}
```

and then `self.add_message("similar-string", **error_message)`.

**Review by atodorov — CHANGES_REQUESTED** [review]
Some changes requested.

**Comment by 17sushmita** [comment]
@atodorov, Could you check why tests have not completed yet?

**Comment by atodorov** [comment]
> @atodorov, Could you check why tests have not completed yet?

That's the latest stupidity from GitHub:
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

Sigh.

**Comment by 17sushmita** [comment]
> > @atodorov, Could you check why tests have not completed yet?
> 
> That's the latest stupidity from GitHub:
> https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
> 
> Sigh.

Oh okay, thanks! I've made the changes.

File: kiwi_lint/similar_string.py (hunk 28 above)
Comment (atodorov) [reply kiwi_lint/similar_string.py:48]: Still unresolved, you've changed the places of the 2 for loops but didn't address the original comment.

**Review by atodorov — CHANGES_REQUESTED** [review]
Minor change needed, see comments. 

However I'm realizing we should have a mechanism to silence some warnings. In Python files that's easy with `# pylint: disable` comments, however we can't do this easily in HTML. 

For example the following errors have similarity but in fact are different things:
```
tcms/templates/navbar.html:98:137: R5011: "Translation guide" is similar to "Translation mode". Try keeping them the same to reduce work for translators! (similar-string)

tcms/bugs/templates/bugs/get.html:131:42: R5011: "Close" is similar to "Clone". Try keeping them the same to reduce work for translators! (similar-string)
```

What's the normalized distance for them? Can you adjust the error message to say something like `"Close" is xy% similar to "Clone"` so we can have this information at hand ?

File: kiwi_lint/similar_string.py (hunk 28 above)
Comment (17sushmita) [reply kiwi_lint/similar_string.py:48]: Actually, I wanted to only detect .txt files in `templates/email` folder because other txt files are not template files for ex - there are required dependency .txt files in `requirements` folder and there is .txt file in `tests/bugzilla` folder too. Although there won't be any such tag like trans and blocktrans in them but I think we shouldn't unnecessarily scan them if they aren't template files. And after making the change i.e changing placement of the loop, we're detecting that if the current root is `templates/email` then only we'll add txt files to the `all_template_files` set. Previously, there was another loop other than previous loop to only parse through `template/email` folder.

**Comment by 17sushmita** [comment]
@atodorov, could you please approve the workflows again?

File: kiwi_lint/similar_string.py (hunk 28 above)
Comment (atodorov) [reply kiwi_lint/similar_string.py:49]: FTR other .txt files will not have the trans or blocktrans tags inside of them so they shouldn't affect the results.
